src/batch/src/executor/join/lookup_join.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::array::DataChunk;
+use risingwave_common::catalog::Schema;
+use risingwave_common::error::{ErrorCode, Result, RwError};
+use risingwave_common::types::DataType;
+use risingwave_common::util::chunk_coalesce::{DataChunkBuilder, SlicedDataChunk};
+use risingwave_expr::expr::{build_from_prost, BoxedExpression};
+use risingwave_pb::batch_plan::plan_node::NodeBody;
+
+use crate::executor::join::row_level_iter::RowLevelIter;
+use crate::executor::join::{
+    concatenate, convert_datum_refs_to_chunk, convert_row_to_chunk, JoinType,
+};
+use crate::executor::{
+    BoxedDataChunkStream, BoxedExecutor, BoxedExecutorBuilder, Executor, ExecutorBuilder,
+};
+use crate::task::BatchTaskContext;
+
+pub struct LookupJoinExecutor {
+    join_type: JoinType,
+    condition: Option<BoxedExpression>,
+    probe_side_source: RowLevelIter,
+    probe_side_schema: Vec<DataType>,
+    probe_side_idxs: Vec<usize>,
+    build_side: RowLevelIter,
+    build_side_schema: Vec<DataType>,
+    build_side_idxs: Vec<usize>,
+    chunk_builder: DataChunkBuilder,
+    schema: Schema,
+    output_indices: Vec<usize>,
+    last_chunk: Option<SlicedDataChunk>,
+    identity: String,
+}
+
+impl Executor for LookupJoinExecutor {
+    fn schema(&self) -> &Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl LookupJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        loop {
+            let cur_row = self.probe_side_source.get_current_row_ref();
+            if cur_row.is_some() {
+                // TODO: lookup rows on build side table using index scan, with cur_probe_row as
+                // input
```

Comment:
We don't need index scan here. The lookup join process is sending `RowSeqScanNode` to compute node to fetch result. It's quite similar to it: https://github.com/singularity-data/risingwave/blob/9abfc35b3c678cecef6260cf52fa6964cbe58a47/src/batch/src/execution/grpc_exchange.rs#L43

But you need to construct `RowSeqScanNode` using probe side data.